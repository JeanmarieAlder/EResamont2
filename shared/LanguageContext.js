import { createContext } from "react";

export const LanguageContext = createContext();
// 1 is French
// 2 is Italian
// 3 is English
// 4 is Deutsch
