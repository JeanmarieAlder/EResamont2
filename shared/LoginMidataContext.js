import React, { useState, useEffect, useContext } from "react";
import { ToastAndroid } from "react-native";
import * as AppAuth from "expo-app-auth";
import * as SecureStore from "expo-secure-store";
import requestGetMidata from "../utils/requestGetMidata";

export const Auth2Context = React.createContext();
export const useAuth2 = () => useContext(Auth2Context);
export const Auth2Provider = ({ children }) => {
  const StorageKey = "EResamont2MidataOAuthKey";
  const config = {
    issuer: "https://test.midata.coop/fhir/metadata",
    clientId: "eresamont2-test",
    serviceConfiguration: {
      authorizationEndpoint: "https://test.midata.coop/authservice",
      tokenEndpoint: "https://test.midata.coop/v1/token",
      revocationEndpoint: "https://test.midata.coop/v1/token"
    },
    redirectUrl: AppAuth.OAuthRedirect + ":/oauthredirect"
  };

  const [authState, setAuthState] = useState(null);
  const [userName, setUserName] = useState(null);

  useEffect(() => {
    console.log("LoginMidataContext: useEffect()");
    (async () => {
      let cachedAuth = await getCachedAuthAsync();
      if (cachedAuth && !authState) {
        setAuthState(cachedAuth);
      }
    })();
  }, []);

  const signInAsync = async () => {
    console.log("signInAsync()");
    let authState = await AppAuth.authAsync(config);
    await cacheAuthAsync(authState);
    console.log("signInAsync", authState);
    setAuthState(authState);
    return authState;
  };

  const signOutAsync = async ({ accessToken }) => {
    try {
      await AppAuth.revokeAsync(config, {
        token: accessToken,
        isClientIdProvided: true
      });
      await SecureStore.deleteItemAsync(StorageKey);
      console.log("sign out completed");
      ToastAndroid.show("Sign out completed", ToastAndroid.SHORT);
      setAuthState(null);
      return null;
    } catch (e) {
      console.error(`Failed to revoke token: ${e.message}`);
    }
  };

  const cacheAuthAsync = async authState => {
    return await SecureStore.setItemAsync(
      StorageKey,
      JSON.stringify(authState)
    );
  };

  const getCachedAuthAsync = async () => {
    let value = await SecureStore.getItemAsync(StorageKey);
    let authState = JSON.parse(value);
    console.log("getCachedAuthAsync", authState);
    if (authState) {
      if (checkIfTokenExpired(authState)) {
        return refreshAuthAsync(authState);
      } else {
        return authState;
      }
    }
    return null;
  };

  const checkIfTokenExpired = ({ accessTokenExpirationDate }) => {
    return new Date(accessTokenExpirationDate) < new Date();
  };

  const refreshAuthAsync = async ({ refreshToken }) => {
    let authState = await AppAuth.refreshAsync(config, refreshToken);
    console.log("refreshAuth", authState);
    await cacheAuthAsync(authState);
    return authState;
  };

  const getLoggedUserName = async () => {
    if (authState) {
      let response = await requestGetMidata(
        "https://test.midata.coop/fhir/Patient/" +
          authState.additionalParameters.patient,
        authState,
        signInAsync
      );
      if (response) {
        setUserName(response.name[0].given[0]);
        return true;
      }
    }
    return false;
  };

  return (
    <Auth2Context.Provider
      value={{
        authState,
        userName,
        signInAsync,
        signOutAsync,
        cacheAuthAsync,
        getCachedAuthAsync,
        checkIfTokenExpired,
        refreshAuthAsync,
        getLoggedUserName
      }}
    >
      {children}
    </Auth2Context.Provider>
  );
};
