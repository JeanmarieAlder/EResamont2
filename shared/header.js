import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { MaterialIcons } from "@expo/vector-icons";

export default function Header({ title, navigation }) {
  const openMenu = () => {
    navigation.openDrawer();
  };

  return (
    <View style={styles.header}>
      <MaterialIcons
        name="menu"
        size={28}
        onPress={openMenu}
        style={styles.icon}
      />
      <View>
        <Text style={styles.headerText}>{title}</Text>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  header: {
    flexDirection: "row",
    justifyContent: "flex-end",
    flex: 1
  },
  headerText: {
    fontWeight: "bold",
    fontSize: 20,
    color: "#fff",
    alignItems: "flex-end",
    letterSpacing: 1,
    marginLeft: 10
  },
  icon: { color: "#fff" }
});
