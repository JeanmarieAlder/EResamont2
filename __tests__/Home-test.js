import * as React from "react";
import renderer from "react-test-renderer";
import { LanguageContext } from "../shared/LanguageContext";

import Home from "../screens/Home";

describe("Home", () => {
  const language = 1; //French

  it(`renders correctly without content`, () => {
    const tree = renderer
      .create(
        <LanguageContext.Provider value={{ language }}>
          <Home />
        </LanguageContext.Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
