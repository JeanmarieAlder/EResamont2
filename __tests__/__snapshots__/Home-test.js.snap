// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly without content 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": "100%",
          "width": "100%",
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "marginBottom": 25,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": "100%",
            "resizeMode": "contain",
            "width": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <RCTScrollView
        contentContainerStyle={
          Object {
            "alignItems": "center",
            "flexGrow": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <View>
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "darkblue",
                "borderRadius": 10,
                "height": 40,
                "justifyContent": "center",
                "marginTop": 10,
                "opacity": 1,
                "padding": 10,
                "width": "80%",
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontFamily": "sans-serif-medium",
                  "textAlign": "center",
                  "textAlignVertical": "center",
                }
              }
            >
              Refresh (debug)
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <ActivityIndicator
              animating={true}
              color="black"
              hidesWhenStopped={true}
              size="large"
            />
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
